Bug 1258861 - Store text log step data + errors in main database

wlach opened the pull request:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/mozilla/treeherder/1696)
<!-- Reviewable:end -->

indygreg commented on treeherder/model/models.py:
```diff
@@ -1137,6 +1137,70 @@ def __str__(self):
+    name = models.CharField(max_length=200)
+    started = models.DateTimeField(null=True)
+    finished = models.DateTimeField(null=True)
+    started_line_number = models.PositiveIntegerField()
```
If we store the byte offset instead of or in addition to the line number, we could facilitate clients sending HTTP Range requests to request subsets of logs. This allows clients to get the logs for a step without having to download the log and parse newlines.

wlach replied:
This is a good idea, let's do it in a followup though.

jgraham replied:
Does that work given that the resources are compressed?

camd commented on tests/autoclassify/utils.py:
```diff
@@ -66,35 +71,30 @@ def get_data(base_data, updates):
         if not data:
             continue
+        error = TextLogError.objects.create(job=job, step=step,
+                                            line=formatter(data).split("\n")[0],
```
you could also use ``rstrip()`` to remove the newlines or any trailing whitespace.

wlach replied:
Not sure if this will work if there are multiple lines? I'm basing this off existing code, so I'm a bit reluctant to change logic that we know works.

camd commented on treeherder/log_parser/management/commands/crossreference_error_lines.py:
```diff
@@ -96,20 +88,20 @@ def crossreference_error_lines(self, repository, job_guid, failure_lines, text_l
 
         # For each error in the text log summary artifact, try to match the next
```
nit, but the comment here should be adjusted since there's no more artifact.  :)

wlach replied:
Right :)

camd commented:
Reviewed 3 of 31 files at r1, 1 of 8 files at r4, 20 of 44 files at r6.
Review status: 24 of 34 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[ui/js/controllers/logviewer.js, line 283 \[r6\]](https://reviewable.io:443/reviews/mozilla/treeherder/1696#-KQ7l2AyHW2BVin5h2CC:-KQ7l2AzqLR6Ob-mjJsa:b-a9evp4) ([raw file](https://github.com/mozilla/treeherder/blob/29621573e67caef16db9fde245f2d55e13ca21af/ui/js/controllers/logviewer.js#L283)):*
> ```JavaScript
>                 // load the first failure step line in the artifact. We
>                 // also need to test for the 0th element for outlier jobs.
>                 var allErrors = _.flatten(textLogSteps.map(s => s.errors));
> ```

Nice use of the new lambda!  :)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mozilla/treeherder/1696)*
<!-- Sent from Reviewable.io -->

camd commented:
Review status: 24 of 34 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[treeherder/model/derived/artifacts.py, line 147 \[r6\]](https://reviewable.io:443/reviews/mozilla/treeherder/1696#-KQ7nZeMzyR1fi4b-GNE:-KQ7nZeNSyOB-DUF7CQH:bsjn9e5) ([raw file](https://github.com/mozilla/treeherder/blob/29621573e67caef16db9fde245f2d55e13ca21af/treeherder/model/derived/artifacts.py#L147)):*
> ```Python
>                 if step.get(tkey):
>                     defaults[tkey] = dateutil.parser.parse(
>                         step[tkey], ignoretz=True)
> ```

are we sure we want to ``ignoretz`` here?  Would you just be up for adding a comment why we're doing this?  I can't remember if these times always come in as UTC or not.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mozilla/treeherder/1696)*
<!-- Sent from Reviewable.io -->

wlach commented:
Review status: 0 of 35 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[treeherder/model/derived/artifacts.py, line 147 \[r6\]](https://reviewable.io:443/reviews/mozilla/treeherder/1696#-KQ7nZeMzyR1fi4b-GNE:-KQMmG5wx51EZpbgSf2Q:b-ewthpi) ([raw file](https://github.com/mozilla/treeherder/blob/29621573e67caef16db9fde245f2d55e13ca21af/treeherder/model/derived/artifacts.py#L147)):*
<details><summary><i>Previously, camd wrote…</i></summary>
> are we sure we want to ``ignoretz`` here?  Would you just be up for adding a comment why we're doing this?  I can't remember if these times always come in as UTC or not.
</details>
So dateutil.parse will choke if it encounters timezone information, since we haven't configured django to support it (see treeherder/etl/bugzilla.py). I believe the log steps that we produce don't have timezone info, but I think it makes sense to do things this way just in case. I'll add a comment though.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mozilla/treeherder/1696)*
<!-- Sent from Reviewable.io -->

jgraham commented:
Reviewed 1 of 37 files at r7, 28 of 32 files at r8.
Review status: 29 of 35 files reviewed at latest revision, 20 unresolved discussions, some commit checks failed.

---

*[treeherder/log_parser/management/commands/crossreference_error_lines.py, line 59 \[r8\]](https://reviewable.io:443/reviews/mozilla/treeherder/1696#-KQaDcD-m-MkaEJSh_Sp:-KQaDcD0Cp33IkHA6ROi:b-uhn7zy) ([raw file](https://github.com/mozilla/treeherder/blob/5b203119dec627605cb7524bfc2765467c72c758/treeherder/log_parser/management/commands/crossreference_error_lines.py#L59)):*
> ```Python
>             bug_suggestions = bug_suggestions[0]
> 
>         self.crossreference_error_lines(repository, job_guid, failure_lines,
> ```

Reformat this to one argument per line.

---

*[treeherder/log_parser/management/commands/crossreference_error_lines.py, line 60 \[r8\]](https://reviewable.io:443/reviews/mozilla/treeherder/1696#-KQaDpztjV_V0yvKwNmh:-KQaDpztjV_V0yvKwNmi:b-h7lpkv) ([raw file](https://github.com/mozilla/treeherder/blob/5b203119dec627605cb7524bfc2765467c72c758/treeherder/log_parser/management/commands/crossreference_error_lines.py#L60)):*
> ```Python
> 
>         self.crossreference_error_lines(repository, job_guid, failure_lines,
>                                         TextLogError.objects.filter(job__guid=job_guid),
> ```

This probably wants to specify that it's ordered by `line_number`.

---

*[treeherder/log_parser/management/commands/crossreference_error_lines.py, line 92 \[r8\]](https://reviewable.io:443/reviews/mozilla/treeherder/1696#-KQaEVdF8a_N7rH2q1fD:-KQaEVdG6sG1rWbA_Zlw:bgakh6n) ([raw file](https://github.com/mozilla/treeherder/blob/5b203119dec627605cb7524bfc2765467c72c758/treeherder/log_parser/management/commands/crossreference_error_lines.py#L92)):*
> ```Python
>         # structured log line
>         for error in text_log_errors:
>             if regexp and regexp.match(error.line):
> ```

Why is it sae to remove the `.strip()`?

---

*[treeherder/model/models.py, line 1139 \[r8\]](https://reviewable.io:443/reviews/mozilla/treeherder/1696#-KQ_hvgh1XQJzbbwF1dk:-KQ_hvgktZmYbKHz-I80:bndw2te) ([raw file](https://github.com/mozilla/treeherder/blob/5b203119dec627605cb7524bfc2765467c72c758/treeherder/model/models.py#L1139)):*
> ```Python
> 
> class TextLogStep(models.Model):
>     '''
> ```

This may be irrelevant, but I notice that all your code uses single quotes for docstrings and all the existing code uses double quotes.

---

*[treeherder/model/models.py, line 1187 \[r8\]](https://reviewable.io:443/reviews/mozilla/treeherder/1696#-KQ_i7oxbCzVxayIWRqt:-KQ_i7oyJX1ZgnUtBqCN:b-6zyr3t) ([raw file](https://github.com/mozilla/treeherder/blob/5b203119dec627605cb7524bfc2765467c72c758/treeherder/model/models.py#L1187)):*
> ```Python
>     job = FlexibleForeignKey(Job)
>     step = FlexibleForeignKey(TextLogStep, related_name='errors')
>     line = models.CharField(max_length=1024)
> ```

Where does 1024 come from and are we happy with it?

---

*[treeherder/model/derived/artifacts.py, line 129 \[r8\]](https://reviewable.io:443/reviews/mozilla/treeherder/1696#-KQa1u-278y6ugRJjAVS:-KQa1u-3RC6eVXcbvHDX:baatf5q) ([raw file](https://github.com/mozilla/treeherder/blob/5b203119dec627605cb7524bfc2765467c72c758/treeherder/model/derived/artifacts.py#L129)):*
> ```Python
>                 **job_detail_dict)
> 
>     def store_text_log_summary(self, job, text_log_summary_artifact):
> ```

This doesn't really seem like an instance method. Any reason it isn't a classmethod or a standalone function?

---

*[treeherder/model/derived/artifacts.py, line 135 \[r8\]](https://reviewable.io:443/reviews/mozilla/treeherder/1696#-KQa22vrMevUS44kkORv:-KQa22vsqTdxEMWKqp-v:bykaybq) ([raw file](https://github.com/mozilla/treeherder/blob/5b203119dec627605cb7524bfc2765467c72c758/treeherder/model/derived/artifacts.py#L135)):*
> ```Python
>         step_data = json.loads(
>             text_log_summary_artifact['blob'])['step_data']
>         result_map = dict((v, k) for (k, v) in TextLogStep.RESULTS)
> ```

Use a dictionary literal i.e. `{value: key for (key,value) in TextLogStep.RESULTS}`

---

*[treeherder/model/derived/artifacts.py, line 153 \[r8\]](https://reviewable.io:443/reviews/mozilla/treeherder/1696#-KQa3muRi4f7IlzED89-:-KQa3muSxy9NyL20P4Xo:b-eagszx) ([raw file](https://github.com/mozilla/treeherder/blob/5b203119dec627605cb7524bfc2765467c72c758/treeherder/model/derived/artifacts.py#L153)):*
> ```Python
>                         step[tkey], ignoretz=True)
> 
>             log_step, _ = TextLogStep.objects.get_or_create(
> ```

Are we OK with not updating the information if it differs? Or checking that it doesn't differ?

---

*[treeherder/model/derived/artifacts.py, line 162 \[r8\]](https://reviewable.io:443/reviews/mozilla/treeherder/1696#-KQa3wOknEecS1it_Z9M:-KQa3wOknEecS1it_Z9N:ba8skln) ([raw file](https://github.com/mozilla/treeherder/blob/5b203119dec627605cb7524bfc2765467c72c758/treeherder/model/derived/artifacts.py#L162)):*
> ```Python
>                 for error in step['errors']:
>                     _, error_created = TextLogError.objects.get_or_create(
>                         job=job, line_number=error['linenumber'],
> ```

newline after `,` please

---

*[treeherder/webapp/api/serializers.py, line 162 \[r8\]](https://reviewable.io:443/reviews/mozilla/treeherder/1696#-KQaFipHIdXhmQiWEin6:-KQaFipHIdXhmQiWEin7:b-mz5sov) ([raw file](https://github.com/mozilla/treeherder/blob/5b203119dec627605cb7524bfc2765467c72c758/treeherder/webapp/api/serializers.py#L162)):*
> ```Python
>     class Meta:
>         model = models.TextLogError
>         exclude = ['id', 'job', 'step']
> ```

Excluding the id seems pretty weird? I don't exactly know why you would want to look up an error line by id, but it at least seems plausible.

---

*[treeherder/webapp/api/serializers.py, line 171 \[r8\]](https://reviewable.io:443/reviews/mozilla/treeherder/1696#-KQaTqcAX9KdbXjkKaAm:-KQaTqcCyF-W24yatRwx:bhkipi4) ([raw file](https://github.com/mozilla/treeherder/blob/5b203119dec627605cb7524bfc2765467c72c758/treeherder/webapp/api/serializers.py#L171)):*
> ```Python
> 
>     def get_result(self, obj):
>         return obj.get_result_display()
> ```

So much magic :/

---

*[treeherder/webapp/api/serializers.py, line 175 \[r8\]](https://reviewable.io:443/reviews/mozilla/treeherder/1696#-KQaGNJHGx3fNbDQ5JYA:-KQaGNJHGx3fNbDQ5JYB:b-veie6u) ([raw file](https://github.com/mozilla/treeherder/blob/5b203119dec627605cb7524bfc2765467c72c758/treeherder/webapp/api/serializers.py#L175)):*
> ```Python
>     class Meta:
>         model = models.TextLogStep
>         exclude = ['id', 'job']
> ```

Again, I'm not sure we should exclude `id`.

---

*[ui/js/controllers/logviewer.js, line 38 \[r8\]](https://reviewable.io:443/reviews/mozilla/treeherder/1696#-KQaUTarIR46DFkwc7-I:-KQaUTasaOwVPcaHWbW1:b-1tnsix) ([raw file](https://github.com/mozilla/treeherder/blob/5b203119dec627605cb7524bfc2765467c72c758/ui/js/controllers/logviewer.js#L38)):*
> ```JavaScript
> 
>         $scope.$watch('steps', function () {
>             if (!$scope.steps) {
> ```

I think I find this use of an initially implicitly undefined $scope.steps to be quite confusing.

---

*[ui/js/controllers/logviewer.js, line 276 \[r8\]](https://reviewable.io:443/reviews/mozilla/treeherder/1696#-KQaUqS_FvZh73agU0d5:-KQaUqSapB6nyaK9yrki:bi3w9t5) ([raw file](https://github.com/mozilla/treeherder/blob/5b203119dec627605cb7524bfc2765467c72c758/ui/js/controllers/logviewer.js#L276)):*
> ```JavaScript
>                 // add an ordering to each step
>                 var i = 0;
>                 $scope.steps.forEach(function(step) {
> ```

The `i` variable is unneeded, just write

    textLogSteps.forEach((step, i) => {step.order = i});

---

*[ui/plugins/failure_summary/controller.js, line 78 \[r8\]](https://reviewable.io:443/reviews/mozilla/treeherder/1696#-KQaKeZeGSLOAciuWT8g:-KQaKeZgLJjPtfvhePvk:b-crwjss) ([raw file](https://github.com/mozilla/treeherder/blob/5b203119dec627605cb7524bfc2765467c72c758/ui/plugins/failure_summary/controller.js#L78)):*
> ```JavaScript
>                                 project: $rootScope.repoName,
>                                 jobId: newValue
>                             }, function(textLogSteps) {
> ```

/me wonders who designed this to not be a promise.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mozilla/treeherder/1696)*
<!-- Sent from Reviewable.io -->

jgraham commented:
Reviewed 2 of 37 files at r7, 4 of 32 files at r8.
Review status: all files reviewed at latest revision, 22 unresolved discussions, some commit checks failed.

---

*[treeherder/model/management/commands/migrate_text_log_artifacts.py, line 85 \[r8\]](https://reviewable.io:443/reviews/mozilla/treeherder/1696#-KQaebd3ctblGQ53GjJi:-KQaebd4xDFH4YPk_g2h:b-da8ph8) ([raw file](https://github.com/mozilla/treeherder/blob/5b203119dec627605cb7524bfc2765467c72c758/treeherder/model/management/commands/migrate_text_log_artifacts.py#L85)):*
> ```Python
>                         'job__project_specific_id', flat=True))
>                 start = time.time()
>                 if ds_job_ids:
> ```

Guess it would be nice if we could avoid this giant indented block

---

*[treeherder/model/management/commands/migrate_text_log_artifacts.py, line 93 \[r8\]](https://reviewable.io:443/reviews/mozilla/treeherder/1696#-KQaf4bqN7YHnEFCq35t:-KQaf4brNvKDRULf6uh3:b-3gikw5) ([raw file](https://github.com/mozilla/treeherder/blob/5b203119dec627605cb7524bfc2765467c72c758/treeherder/model/management/commands/migrate_text_log_artifacts.py#L93)):*
> ```Python
>                     text_log_errors = []
> 
>                     def unwrap(row):
> ```

It seems like this could be a standalone function since it doesn't close over anything.

---

*[treeherder/model/management/commands/migrate_text_log_artifacts.py, line 104 \[r8\]](https://reviewable.io:443/reviews/mozilla/treeherder/1696#-KQafP6oVJUtRtfR-fuW:-KQafP6pUBTc41_zixJQ:bgt33b6) ([raw file](https://github.com/mozilla/treeherder/blob/5b203119dec627605cb7524bfc2765467c72c758/treeherder/model/management/commands/migrate_text_log_artifacts.py#L104)):*
> ```Python
> 
>                     with transaction.atomic():
>                         with concurrent.futures.ThreadPoolExecutor(max_workers=8) as executor:
> ```

Doesn't this require python 3? Or are you installing a backport in a one-off way?

---

*[treeherder/model/migrations/0038_text_log_db.py, line 43 \[r8\]](https://reviewable.io:443/reviews/mozilla/treeherder/1696#-KQaaeuRQucQIQ8Mqdx7:-KQaaeuSta-SB3I1BLne:b-hkkrsa) ([raw file](https://github.com/mozilla/treeherder/blob/5b203119dec627605cb7524bfc2765467c72c758/treeherder/model/migrations/0038_text_log_db.py#L43)):*
> ```Python
>             },
>         ),
>         migrations.AddField(
> ```

Was this all autogenrerated? It's unclear to me why this ended up in a separate migration step.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mozilla/treeherder/1696)*
<!-- Sent from Reviewable.io -->

wlach commented:
Review status: all files reviewed at latest revision, 22 unresolved discussions, some commit checks failed.

---

*[treeherder/log_parser/management/commands/crossreference_error_lines.py, line 60 \[r8\]](https://reviewable.io:443/reviews/mozilla/treeherder/1696#-KQaDpztjV_V0yvKwNmh:-KR5By-DQgIhi8FaInZw:bp7xbw9) ([raw file](https://github.com/mozilla/treeherder/blob/5b203119dec627605cb7524bfc2765467c72c758/treeherder/log_parser/management/commands/crossreference_error_lines.py#L60)):*
<details><summary><i>Previously, jgraham wrote…</i></summary>
> This probably wants to specify that it's ordered by `line_number`.
</details>
How do you mean?

---

*[treeherder/log_parser/management/commands/crossreference_error_lines.py, line 92 \[r8\]](https://reviewable.io:443/reviews/mozilla/treeherder/1696#-KQaEVdF8a_N7rH2q1fD:-KR5C5S6jsn1F0TpshD2:b-itmtf2) ([raw file](https://github.com/mozilla/treeherder/blob/5b203119dec627605cb7524bfc2765467c72c758/treeherder/log_parser/management/commands/crossreference_error_lines.py#L92)):*
<details><summary><i>Previously, jgraham wrote…</i></summary>
> Why is it sae to remove the `.strip()`?
</details>
Hmm, I think the removal was accidental. I'll just add it back in.

---

*[treeherder/model/models.py, line 1171 \[r5\]](https://reviewable.io:443/reviews/mozilla/treeherder/1696#-KMkU3owPGqroJzazBxS-r5-1171:-KR5-4da7p9Tamiefcp5:bltywe4) ([raw file](https://github.com/mozilla/treeherder/blob/3f2928bd6ee986bddfa342018a5c804cfe13bdd2/treeherder/model/models.py#L1171)):*
<details><summary><i>Previously, jgraham wrote…</i></summary>
> Does that work given that the resources are compressed?
</details>
Probably not, good point

---

*[treeherder/model/models.py, line 1139 \[r8\]](https://reviewable.io:443/reviews/mozilla/treeherder/1696#-KQ_hvgh1XQJzbbwF1dk:-KR4zqPBNoLUzUKfhOFu:b-n0l7vm) ([raw file](https://github.com/mozilla/treeherder/blob/5b203119dec627605cb7524bfc2765467c72c758/treeherder/model/models.py#L1139)):*
<details><summary><i>Previously, jgraham wrote…</i></summary>
> This may be irrelevant, but I notice that all your code uses single quotes for docstrings and all the existing code uses double quotes.
</details>
May as well be consistent. I'll fix.

---

*[treeherder/model/models.py, line 1187 \[r8\]](https://reviewable.io:443/reviews/mozilla/treeherder/1696#-KQ_i7oxbCzVxayIWRqt:-KR5-Cc2mCjj5f_f6IzZ:b-35ugg1) ([raw file](https://github.com/mozilla/treeherder/blob/5b203119dec627605cb7524bfc2765467c72c758/treeherder/model/models.py#L1187)):*
<details><summary><i>Previously, jgraham wrote…</i></summary>
> Where does 1024 come from and are we happy with it?
</details>
As discussed in the meeting, I'll just remove this restriction for now (since the old text log artifacts had no such restrictions, nor do the existing failure line models in the db)

---

*[treeherder/model/management/commands/migrate_text_log_artifacts.py, line 85 \[r8\]](https://reviewable.io:443/reviews/mozilla/treeherder/1696#-KQaebd3ctblGQ53GjJi:-KR55wuCvM5Wo5ZNFPGE:b-jo9f0r) ([raw file](https://github.com/mozilla/treeherder/blob/5b203119dec627605cb7524bfc2765467c72c758/treeherder/model/management/commands/migrate_text_log_artifacts.py#L85)):*
<details><summary><i>Previously, jgraham wrote…</i></summary>
> Guess it would be nice if we could avoid this giant indented block
</details>
If you have any suggestions, let me know. Remember that this is just a one-off migration script which we'll be deleting shortly after this lands, so it's probably not worth bikeshedding over too much.

---

*[treeherder/model/management/commands/migrate_text_log_artifacts.py, line 93 \[r8\]](https://reviewable.io:443/reviews/mozilla/treeherder/1696#-KQaf4bqN7YHnEFCq35t:-KR55r7PY6ro-KlXNZbB:b6m6rco) ([raw file](https://github.com/mozilla/treeherder/blob/5b203119dec627605cb7524bfc2765467c72c758/treeherder/model/management/commands/migrate_text_log_artifacts.py#L93)):*
<details><summary><i>Previously, jgraham wrote…</i></summary>
> It seems like this could be a standalone function since it doesn't close over anything.
</details>
It could, I think it's better to put it near where it's used though.

---

*[treeherder/model/management/commands/migrate_text_log_artifacts.py, line 104 \[r8\]](https://reviewable.io:443/reviews/mozilla/treeherder/1696#-KQafP6oVJUtRtfR-fuW:-KR55j9lSauGhkLTRYcC:b-vl4imi) ([raw file](https://github.com/mozilla/treeherder/blob/5b203119dec627605cb7524bfc2765467c72c758/treeherder/model/management/commands/migrate_text_log_artifacts.py#L104)):*
<details><summary><i>Previously, jgraham wrote…</i></summary>
> Doesn't this require python 3? Or are you installing a backport in a one-off way?
</details>
We have concurrent futures installed as a library for cases like this. :)

---

*[treeherder/model/migrations/0038_text_log_db.py, line 43 \[r8\]](https://reviewable.io:443/reviews/mozilla/treeherder/1696#-KQaaeuRQucQIQ8Mqdx7:-KR56BJasBpqXOAKsX5e:b7oy6h7) ([raw file](https://github.com/mozilla/treeherder/blob/5b203119dec627605cb7524bfc2765467c72c758/treeherder/model/migrations/0038_text_log_db.py#L43)):*
<details><summary><i>Previously, jgraham wrote…</i></summary>
> Was this all autogenrerated? It's unclear to me why this ended up in a separate migration step.
</details>
Yeah, it's autogenerated.

---

*[treeherder/webapp/api/serializers.py, line 162 \[r8\]](https://reviewable.io:443/reviews/mozilla/treeherder/1696#-KQaFipHIdXhmQiWEin6:-KR5Ahhwl-RZUSqYaMj2:b-a4ozli) ([raw file](https://github.com/mozilla/treeherder/blob/5b203119dec627605cb7524bfc2765467c72c758/treeherder/webapp/api/serializers.py#L162)):*
<details><summary><i>Previously, jgraham wrote…</i></summary>
> Excluding the id seems pretty weird? I don't exactly know why you would want to look up an error line by id, but it at least seems plausible.
</details>
I can leave it in for now.

---

*[ui/js/controllers/logviewer.js, line 38 \[r8\]](https://reviewable.io:443/reviews/mozilla/treeherder/1696#-KQaUTarIR46DFkwc7-I:-KR5Au0yH3MKRLOE9-TD:b-er1xr2) ([raw file](https://github.com/mozilla/treeherder/blob/5b203119dec627605cb7524bfc2765467c72c758/ui/js/controllers/logviewer.js#L38)):*
<details><summary><i>Previously, jgraham wrote…</i></summary>
> I think I find this use of an initially implicitly undefined $scope.steps to be quite confusing.
</details>
It's basically inline with how the logviewer was originally written, which isn't the best but I don't think this is the best place to do a large-scale refactoring.

---

*[ui/js/controllers/logviewer.js, line 276 \[r8\]](https://reviewable.io:443/reviews/mozilla/treeherder/1696#-KQaUqS_FvZh73agU0d5:-KR5BH7Nalgh3YoIfXBN:b-896fix) ([raw file](https://github.com/mozilla/treeherder/blob/5b203119dec627605cb7524bfc2765467c72c758/ui/js/controllers/logviewer.js#L276)):*
<details><summary><i>Previously, jgraham wrote…</i></summary>
> The `i` variable is unneeded, just write
> 
>     textLogSteps.forEach((step, i) => {step.order = i});
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mozilla/treeherder/1696)*
<!-- Sent from Reviewable.io -->

wlach commented:
Review status: all files reviewed at latest revision, 22 unresolved discussions, some commit checks failed.

---

*[treeherder/model/derived/artifacts.py, line 129 \[r8\]](https://reviewable.io:443/reviews/mozilla/treeherder/1696#-KQa1u-278y6ugRJjAVS:-KR5KESNYDS7fxnQihlY:b-rp2t7r) ([raw file](https://github.com/mozilla/treeherder/blob/5b203119dec627605cb7524bfc2765467c72c758/treeherder/model/derived/artifacts.py#L129)):*
<details><summary><i>Previously, jgraham wrote…</i></summary>
> This doesn't really seem like an instance method. Any reason it isn't a classmethod or a standalone function?
</details>
No, however it's following the style elsewhere in the file. Could we fix this in a followup?

---

*[treeherder/model/derived/artifacts.py, line 153 \[r8\]](https://reviewable.io:443/reviews/mozilla/treeherder/1696#-KQa3muRi4f7IlzED89-:-KR5KlDPaHvOb7M0s1wX:b-wc3vxm) ([raw file](https://github.com/mozilla/treeherder/blob/5b203119dec627605cb7524bfc2765467c72c758/treeherder/model/derived/artifacts.py#L153)):*
<details><summary><i>Previously, jgraham wrote…</i></summary>
> Are we OK with not updating the information if it differs? Or checking that it doesn't differ?
</details>
I believe we should only be calling this twice in the case that we accidentally submit the same artifact twice. I guess we could use either...

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mozilla/treeherder/1696)*
<!-- Sent from Reviewable.io -->

jgraham commented:
Reviewed 15 of 15 files at r9.
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[tests/model/derived/test_artifacts_model.py, line 162 \[r9\]](https://reviewable.io:443/reviews/mozilla/treeherder/1696#-KR8HQWorS7fUVX3mH3R:-KR8HQWorS7fUVX3mH3S:b-zfhxci) ([raw file](https://github.com/mozilla/treeherder/blob/8fd2061fd6c2aeb8e03f0cf72da69c47e83d8857/tests/model/derived/test_artifacts_model.py#L162)):*
> ```Python
> 
>     assert TextLogError.objects.count() == 1
>     assert TextLogError.objects.all()[0].line == "07:51:28  WARNING - \\xe2"
> ```

See other comments about unicode-escape

---

*[treeherder/log_parser/management/commands/crossreference_error_lines.py, line 60 \[r8\]](https://reviewable.io:443/reviews/mozilla/treeherder/1696#-KQaDpztjV_V0yvKwNmh:-KR8BYb2pIaH3EHxQD-D:b-14fx4k) ([raw file](https://github.com/mozilla/treeherder/blob/5b203119dec627605cb7524bfc2765467c72c758/treeherder/log_parser/management/commands/crossreference_error_lines.py#L60)):*
<details><summary><i>Previously, wlach (William Lachance) wrote…</i></summary>
> How do you mean?
</details>
I mean `.order_by('line_number')`. Which probably matches the returned order anyway because it's the insertion order, but I don't know if that's guaranteed.

---

*[treeherder/model/tasks.py, line 144 \[r8\]](https://reviewable.io:443/reviews/mozilla/treeherder/1696#-KR8CIc1dFgH9qAoLLEo:-KR8CIc2Keh2TZf33HxG:b-73svzv) ([raw file](https://github.com/mozilla/treeherder/blob/5b203119dec627605cb7524bfc2765467c72c758/treeherder/model/tasks.py#L144)):*
> ```Python
> 
> 
> @task(name='populate-error-summary')
> ```

Is there also a queue that can be removed from the settings?

---

*[treeherder/model/derived/artifacts.py, line 129 \[r8\]](https://reviewable.io:443/reviews/mozilla/treeherder/1696#-KQa1u-278y6ugRJjAVS:-KR8CRpq2CadmePp2FFU:bb74ce2) ([raw file](https://github.com/mozilla/treeherder/blob/5b203119dec627605cb7524bfc2765467c72c758/treeherder/model/derived/artifacts.py#L129)):*
<details><summary><i>Previously, wlach (William Lachance) wrote…</i></summary>
> No, however it's following the style elsewhere in the file. Could we fix this in a followup?
</details>
Sure

---

*[treeherder/model/derived/artifacts.py, line 153 \[r8\]](https://reviewable.io:443/reviews/mozilla/treeherder/1696#-KQa3muRi4f7IlzED89-:-KR8CeUmIhq41tfvAaFu:b-8v930h) ([raw file](https://github.com/mozilla/treeherder/blob/5b203119dec627605cb7524bfc2765467c72c758/treeherder/model/derived/artifacts.py#L153)):*
<details><summary><i>Previously, wlach (William Lachance) wrote…</i></summary>
> I believe we should only be calling this twice in the case that we accidentally submit the same artifact twice. I guess we could use either...
</details>
To me either replacing or erroring seems more natural than silently ignoring the new data, but I don't know how important this is to get right.

---

*[treeherder/model/derived/artifacts.py, line 164 \[r9\]](https://reviewable.io:443/reviews/mozilla/treeherder/1696#-KR8D8WgvxfC3m3WIqPj:-KR8D8WgvxfC3m3WIqPk:bwfcbbt) ([raw file](https://github.com/mozilla/treeherder/blob/8fd2061fd6c2aeb8e03f0cf72da69c47e83d8857/treeherder/model/derived/artifacts.py#L164)):*
> ```Python
>                         line_number=error['linenumber'],
>                         defaults={
>                             'line': error['line'].encode('unicode-escape')
> ```

What problem are we solving here? Encoding e.g. `"ö"` as `"\xf6"` seems unnecessary and confusing. If the issue is just with non-BMP characters we already have a hack to work around our poorly configured MySQL instance in `log_parser/failureline.py::replace_astral`. And also, apologies for not noticing this before.

---

*[treeherder/model/management/commands/migrate_text_log_artifacts.py, line 85 \[r8\]](https://reviewable.io:443/reviews/mozilla/treeherder/1696#-KQaebd3ctblGQ53GjJi:-KR8Efkzjg78A4MzLkkc:bs4ixp3) ([raw file](https://github.com/mozilla/treeherder/blob/5b203119dec627605cb7524bfc2765467c72c758/treeherder/model/management/commands/migrate_text_log_artifacts.py#L85)):*
<details><summary><i>Previously, wlach (William Lachance) wrote…</i></summary>
> If you have any suggestions, let me know. Remember that this is just a one-off migration script which we'll be deleting shortly after this lands, so it's probably not worth bikeshedding over too much.
</details>
OK, fair enough.

---

*[treeherder/model/management/commands/migrate_text_log_artifacts.py, line 93 \[r8\]](https://reviewable.io:443/reviews/mozilla/treeherder/1696#-KQaf4bqN7YHnEFCq35t:-KR8EkKvUVZZW4qUMHQS:b-lntii5) ([raw file](https://github.com/mozilla/treeherder/blob/5b203119dec627605cb7524bfc2765467c72c758/treeherder/model/management/commands/migrate_text_log_artifacts.py#L93)):*
<details><summary><i>Previously, wlach (William Lachance) wrote…</i></summary>
> It could, I think it's better to put it near where it's used though.
</details>
I tend to disagree (and slightly wonder if there is some small perf effect), but since, as you say, this code is shortlived, it's fine as-is.

---

*[treeherder/model/management/commands/migrate_text_log_artifacts.py, line 123 \[r9\]](https://reviewable.io:443/reviews/mozilla/treeherder/1696#-KR8F7AnnGseRpmKcO55:-KR8F7AnnGseRpmKcO56:bkx61qb) ([raw file](https://github.com/mozilla/treeherder/blob/8fd2061fd6c2aeb8e03f0cf72da69c47e83d8857/treeherder/model/management/commands/migrate_text_log_artifacts.py#L123)):*
> ```Python
>                                                     line_number=line_number,
>                                                     step=text_log_step,
>                                                     line=error['line'].encode('unicode-escape')))
> ```

c.f. previous comment about `unicode.escape`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mozilla/treeherder/1696)*
<!-- Sent from Reviewable.io -->

wlach commented:
Review status: all files reviewed at latest revision, 11 unresolved discussions.

---

*[treeherder/log_parser/management/commands/crossreference_error_lines.py, line 89 \[r6\]](https://reviewable.io:443/reviews/mozilla/treeherder/1696#-KMkU3ovKLIZM4jT7Z30-r6-89:-KR9wpduHoHuKVVxBZXS:b-896fix) ([raw file](https://github.com/mozilla/treeherder/blob/29621573e67caef16db9fde245f2d55e13ca21af/treeherder/log_parser/management/commands/crossreference_error_lines.py#L89)):*
<details><summary><i>Previously, wlach (William Lachance) wrote…</i></summary>
> Right :)
</details>
Done.

---

*[treeherder/log_parser/management/commands/crossreference_error_lines.py, line 60 \[r8\]](https://reviewable.io:443/reviews/mozilla/treeherder/1696#-KQaDpztjV_V0yvKwNmh:-KR9wmpPnkMqLQ46-TDS:bepenrt) ([raw file](https://github.com/mozilla/treeherder/blob/5b203119dec627605cb7524bfc2765467c72c758/treeherder/log_parser/management/commands/crossreference_error_lines.py#L60)):*
<details><summary><i>Previously, jgraham wrote…</i></summary>
> I mean `.order_by('line_number')`. Which probably matches the returned order anyway because it's the insertion order, but I don't know if that's guaranteed.
</details>
Gotcha, will change.

---

*[treeherder/model/tasks.py, line 144 \[r8\]](https://reviewable.io:443/reviews/mozilla/treeherder/1696#-KR8CIc1dFgH9qAoLLEo:-KR9wt5_WLE8DUmy1ozr:b8gc3ox) ([raw file](https://github.com/mozilla/treeherder/blob/5b203119dec627605cb7524bfc2765467c72c758/treeherder/model/tasks.py#L144)):*
<details><summary><i>Previously, jgraham wrote…</i></summary>
> Is there also a queue that can be removed from the settings?
</details>
No, not that I can see. I think these tasks just got dumped into the default one.

---

*[treeherder/model/derived/artifacts.py, line 129 \[r8\]](https://reviewable.io:443/reviews/mozilla/treeherder/1696#-KQa1u-278y6ugRJjAVS:-KR9yKjxIObPpnqB2h6G:b-dweent) ([raw file](https://github.com/mozilla/treeherder/blob/5b203119dec627605cb7524bfc2765467c72c758/treeherder/model/derived/artifacts.py#L129)):*
<details><summary><i>Previously, jgraham wrote…</i></summary>
> Sure
</details>
Filed https://bugzilla.mozilla.org/show_bug.cgi?id=1301477 . I think the best thing to do is actually just to nuke the ArtifactModel from orbit and replace it with a few functions. 

---

*[treeherder/model/derived/artifacts.py, line 153 \[r8\]](https://reviewable.io:443/reviews/mozilla/treeherder/1696#-KQa3muRi4f7IlzED89-:-KR9zRAn8JU5Rr6msUBB:bo2qyxx) ([raw file](https://github.com/mozilla/treeherder/blob/5b203119dec627605cb7524bfc2765467c72c758/treeherder/model/derived/artifacts.py#L153)):*
<details><summary><i>Previously, jgraham wrote…</i></summary>
> To me either replacing or erroring seems more natural than silently ignoring the new data, but I don't know how important this is to get right.
</details>
I'll change it to update_or_create. In general I'd like us to be a bit more deliberate about we handle these kinds of races, but I think that can wait.

---

*[treeherder/model/derived/artifacts.py, line 164 \[r9\]](https://reviewable.io:443/reviews/mozilla/treeherder/1696#-KR8D8WgvxfC3m3WIqPj:-KR9zJ3_1Po64Zp5Hnwj:b5vf533) ([raw file](https://github.com/mozilla/treeherder/blob/8fd2061fd6c2aeb8e03f0cf72da69c47e83d8857/treeherder/model/derived/artifacts.py#L164)):*
<details><summary><i>Previously, jgraham wrote…</i></summary>
> What problem are we solving here? Encoding e.g. `"ö"` as `"\xf6"` seems unnecessary and confusing. If the issue is just with non-BMP characters we already have a hack to work around our poorly configured MySQL instance in `log_parser/failureline.py::replace_astral`. And also, apologies for not noticing this before.
</details>
Ok, I factored out the astral support into a module and started using it here. Seems to work for my unit tests, hopefully that will be the case in production as well.

---

*[treeherder/model/management/commands/migrate_text_log_artifacts.py, line 93 \[r8\]](https://reviewable.io:443/reviews/mozilla/treeherder/1696#-KQaf4bqN7YHnEFCq35t:-KRA-GSt2abE-MwIgt4I:b-9127wj) ([raw file](https://github.com/mozilla/treeherder/blob/5b203119dec627605cb7524bfc2765467c72c758/treeherder/model/management/commands/migrate_text_log_artifacts.py#L93)):*
<details><summary><i>Previously, jgraham wrote…</i></summary>
> I tend to disagree (and slightly wonder if there is some small perf effect), but since, as you say, this code is shortlived, it's fine as-is.
</details>
There would probably be a lot of things I'd change if this was going to stick around. :)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mozilla/treeherder/1696)*
<!-- Sent from Reviewable.io -->

wlach commented:
FYI I added some additional updates to this (beyond what was in the review):

* Make sure that the autoclassify text log summary models still get the legacy text_log_artifact references (in case this needs to be reverted after moving into production)
* Removed some legacy code around determining whether we need to generate bug suggestions.

jgraham commented:
Reviewed 9 of 10 files at r10.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[tests/model/derived/test_artifacts_model.py, line 152 at r10](https://reviewable.io:443/reviews/mozilla/treeherder/1696#-KRTJYflo_G1UaNeU50e:-KRTJYfnanzg76Cioemn:b-oks233) ([raw file](https://github.com/mozilla/treeherder/blob/f90ce477985be1cb624d06488380f036b0360e8a/tests/model/derived/test_artifacts_model.py#L152)):*
> "07:51:29  WARNING - \xf0\x9d\x90\x80"

Could we write this as `u"07:51:29  WARNING - \U0001d400"`? That immediately communicates it's a non-BMP character without having to remember that means a 4 byte UTF8 representation, and decoding that the sequence of escapes is indeed such a representation and not 2-2byte characters or whatever.

---

*[tests/model/derived/test_artifacts_model.py, line 166 at r10](https://reviewable.io:443/reviews/mozilla/treeherder/1696#-KRTK4uJEk1w-3OhMO8m:-KRTK4uJEk1w-3OhMO8n:bi772ww) ([raw file](https://github.com/mozilla/treeherder/blob/f90ce477985be1cb624d06488380f036b0360e8a/tests/model/derived/test_artifacts_model.py#L166)):*
> ```Python
> 
>     assert TextLogError.objects.count() == 2
>     assert TextLogError.objects.get(line_number=1587).line == '07:51:28  WARNING - \xc3\x83'.decode('utf-8')
> ```

Again, the unicode representation here would be easier to read.

---

*[treeherder/model/tasks.py, line 144 at r8](https://reviewable.io:443/reviews/mozilla/treeherder/1696#-KR8CIc1dFgH9qAoLLEo:-KRTPewvuyhAsjnqmTVS:b-ytw1fg) ([raw file](https://github.com/mozilla/treeherder/blob/5b203119dec627605cb7524bfc2765467c72c758/treeherder/model/tasks.py#L144)):*
<details><summary><i>Previously, wlach (William Lachance) wrote…</i></summary>
> No, not that I can see. I think these tasks just got dumped into the default one.
</details>
Looks like it used the `error_summary` queue to me.

---

*[treeherder/model/derived/artifacts.py, line 9 at r10](https://reviewable.io:443/reviews/mozilla/treeherder/1696#-KRTRCaGrDQkwqviwDMG:-KRTRCaGrDQkwqviwDMH:bps1is8) ([raw file](https://github.com/mozilla/treeherder/blob/f90ce477985be1cb624d06488380f036b0360e8a/treeherder/model/derived/artifacts.py#L9)):*
> ```Python
> 
> from treeherder.etl.perf import load_perf_artifacts
> from treeherder.etl.text import astral_filter
> ```

Looks like this file is missing from the review?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mozilla/treeherder/1696)*
<!-- Sent from Reviewable.io -->

wlach commented:
Review status: 31 of 33 files reviewed at latest revision, 8 unresolved discussions.

---

*[tests/model/derived/test_artifacts_model.py, line 152 at r10](https://reviewable.io:443/reviews/mozilla/treeherder/1696#-KRTJYflo_G1UaNeU50e:-KRTvxGRBTp3jD1VUmT1:b-o3werm) ([raw file](https://github.com/mozilla/treeherder/blob/f90ce477985be1cb624d06488380f036b0360e8a/tests/model/derived/test_artifacts_model.py#L152)):*
<details><summary><i>Previously, jgraham wrote…</i></summary>
> > "07:51:29  WARNING - \xf0\x9d\x90\x80"
> 
> Could we write this as `u"07:51:29  WARNING - \U0001d400"`? That immediately communicates it's a non-BMP character without having to remember that means a 4 byte UTF8 representation, and decoding that the sequence of escapes is indeed such a representation and not 2-2byte characters or whatever.
</details>
Sure, did some further cleanup of the test while I was at it.

---

*[treeherder/model/tasks.py, line 144 at r8](https://reviewable.io:443/reviews/mozilla/treeherder/1696#-KR8CIc1dFgH9qAoLLEo:-KRTwMxsQTag1TEDcR1n:b-dersau) ([raw file](https://github.com/mozilla/treeherder/blob/5b203119dec627605cb7524bfc2765467c72c758/treeherder/model/tasks.py#L144)):*
<details><summary><i>Previously, jgraham wrote…</i></summary>
> Looks like it used the `error_summary` queue to me.
</details>
Ok, I'll remove that queue!

---

*[treeherder/model/derived/artifacts.py, line 9 at r10](https://reviewable.io:443/reviews/mozilla/treeherder/1696#-KRTRCaGrDQkwqviwDMG:-KRToWt3KnvSMl3X4oVA:b-9sfc13) ([raw file](https://github.com/mozilla/treeherder/blob/f90ce477985be1cb624d06488380f036b0360e8a/treeherder/model/derived/artifacts.py#L9)):*
<details><summary><i>Previously, jgraham wrote…</i></summary>
> Looks like this file is missing from the review?
</details>
It's a straightforward copypaste from the code you wrote earlier, I went ahead and committed it last week to get the migration started on prod:

https://github.com/mozilla/treeherder/blob/master/treeherder/etl/text.py

If there's any adjustments you'd like to see let me know.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mozilla/treeherder/1696)*
<!-- Sent from Reviewable.io -->

jgraham commented:
Reviewed 4 of 4 files at r11.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/mozilla/treeherder/1696)*
<!-- Sent from Reviewable.io -->
